---
title: "Deconvolution_2024"
author: "Yayuan Zhu"
date: "2024-01-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

...

### Setup

```{r message=FALSE, warning=FALSE}
library(Biobase)
library(fs)
library(magrittr)
library(rstudioapi)
library(Seurat)
library(stringr)
library(tidyverse)

source("functions/functions_R.R")
```

## Description data collection

| Data                                                     | Source                                                                                                                                                                                                                                        | Obtained date (dd/mm/yyyy) |
|-----------------|---------------------------------------|-----------------|
| Bulk RNA-sequencing data                                 | In-house                                                                                                                                                                                                                                      | 16/01/2024                 |
| HGNC complete set                                        | HGNC ([URL](HGNC%20https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/hgnc_complete_set_2024-01-01.txt))                                                                                                                  | 16/01/2024                 |
| Non-coding RNA list                                      | HGNC, filtered on non-coding RNA ([URL](https://www.genenames.org/tools/search/#!/?query=&rows=20&start=0&filter=locus_group:%22Non-coding%20RNA%22))                                                                                         | 16/01/2024                 |
| Single cell RNA-sequencing data                          | In-house                                                                                                                                                                                                                                      | 17/01/2024                 |
| AtheroExpress patient data                               | In-house                                                                                                                                                                                                                                      | 15/03/2024                 |
| Alsaigh single cell RNA-sequencing data with neutrophils | Alsaigh T, Evans D, Frankel D, Torkamani A. Decoding the transcriptome of calcified atherosclerotic plaque at single-cell resolution. Commun Biol. 2022 Oct 12;5(1):1084. doi: 10.1038/s42003-022-04056-7. PMID: 36224302; PMCID: PMC9556750. | 25/03/2024                 |

## AtheroExpress data processing

### Importing files

```{r message=FALSE, warning=FALSE}
bulkData <- read_delim("input/AE_bulk_RNA_batch1.minRib.PC_07042023.txt")
completeHGNC <- read_delim("input/hgnc_complete_set.txt")
ncRNA <- read_delim("input/HGNC_NonCodingRNA.txt")

scData <- subset(read_rds("input/full.43p_10X.integrated.cleaned.archetypes.seurat.RDS"), Tissue != "PBMC")
```

### Filter bulk RNA-sequencing data and single cell RNA-sequencing data

In this section, both the bulk RNA-sequencing (bulk RNA-seq) and single cell RNA-sequencing (scRNA-seq) datasets will be processed to ensure they contain identical sets of genes.

#### Integration gene symbols and filtering non-coding RNA

First, gene symbols will be added to the dataset from the HGNC complete archive. Consequently, genes will be removed of which the Ensembl gene ID is not found in this archive and are non-coding RNA.

```{r}
bulkFilter <- bulkData %>% 
  filter(gene %in% completeHGNC$ensembl_gene_id) %>%
  rename("ensembl_gene_id" = "gene") %>%
  left_join(select(completeHGNC, ensembl_gene_id, symbol), by = "ensembl_gene_id") %>%
  select(ensembl_gene_id, symbol, everything()) %>%
  filter(!(symbol %in% ncRNA$Symbol))
```

The bulk RNA-seq dataset has been reduced from 21835 to 19214 genes.

#### Matching genes

Genes are removed from the scRNA-seq dataset which are not found in the bulk RNA-seq dataset.

```{r}
DefaultAssay(scData) <- "RNA"

toKeep <- intersect(Features(scData), bulkFilter$symbol)

bulkFinal <- bulkFilter %>%
  filter(symbol %in% toKeep) %>%
  as.data.frame() %>%
  column_to_rownames("symbol") %>%
  subset(select = -ensembl_gene_id)

counts <- scData[["RNA"]]$counts

countsKeep <- counts[which(rownames(counts) %in% toKeep),]

scFilter <- subset(scData, features = rownames(countsKeep))
```

Now, both the bulk RNA-seq and scRNA-seq datasets each comprise 18114 genes.

In the scRNA-seq dataset, there were 25 cell types identified. However, the frequency of these cell types found in the dataset vary largely. To ensure reliability and accuracy of the dataset, the cell types with a frequency below an arbitrary threshold will be removed.

```{r}
dfCellTypesRename <- as.data.frame(sort(table(Idents(scFilter)))) %>%
  rename(cellType = Var1) %>%
  mutate(cellType = recode(cellType,
                           "CD14+TREM2-TIMP1+HSPA6+ Lipid-stress Activated Foamy Macrophages" = "Lipid-stress Foamy Macrophages", 
                           "CD14+IL1B+SELL+MX1+ Interferon Activated Inflammatory Monocyte-derived Macrophages" = "Interferon Activated Inflammatory Monocyte-derived Macrophages",
                           "CD14+TREM2-OLR1+NLRP3+ Inflammatory Foamy Macrophages" = "Inflammatory Foamy Macrophages",
                           "CD14+IL1B+SELL+CD16- Antigen-presenting Inflammatory Monocyte-derived Macrophages" = "Antigen-presenting Inflammatory Monocyte-derived Macrophages",
                           "CD14+-IL1B+SELL+CD16+ Migrating Inflammatory Monocyte-derived Macrophages" = "Migrating Inflammatory Monocyte-derived Macrophages",
                           "CD14+TNF+TREM2+FOLR2+ Inflammatory Resident-like Lipid Associated Macrophages" = "Inflammatory Resident-like Lipid Associated Macrophages",
                           "CD14+IL1B+SELL+S100A8+ Migrating Inflammatory Monocyte-derived Macrophages" = "S100A8+ Migrating Inflammatory Monocyte-derived Macrophages"))

threshold <- 50

par(mar = c(20, 6, 4, 2))

ggplot(dfCellTypesRename, aes(x = cellType, y = Freq)) +
  geom_bar(stat = "identity",
           aes(fill = Freq >= threshold)) +
  geom_hline(aes(yintercept = threshold,
                 color = "threshold")) +
  scale_color_manual(name = "",
                     values = c("threshold" = "#E66100")) +
  scale_fill_manual(guide = "none",
                    values = c("#5D3A9B", "darkgrey")) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=60, vjust = 1, hjust = 1, size = 8)) +
  coord_cartesian(ylim = c(0, 750))
```

The plot above, but in a shiny application:

```{r}
# knitr::include_url("https://yyzo.shinyapps.io/Frequence_sequenced_cell_type/",
#                    height = "630px")
```

From the plots above and the previous work of Gemma, it has been decided to keep cell types which have a frequency of 50 and higher.

```{r}
dfCellTypes <- as.data.frame(sort(table(Idents(scFilter)))) %>%
  rename(cellType = Var1) %>%
  mutate(cellType = as.character(cellType))

aboveThreshold <- factor(dfCellTypes$cellType[dfCellTypes$Freq >= 50])

scFinal <- subset(scFilter, idents = aboveThreshold)
```

```{r include=FALSE}
#Remove redundant Seurat objects in order to free up space.
rm(scData, scFilter, bulkData, counts, countsKeep, completeHGNC, ncRNA)
gc()
```

### Clustering

The final scRNA-sequencing dataset contains 21 cell types. How these cell types are clustered can be seen in the UMAP plot below.

```{r}
scRename <- scFinal

ct <- Idents(scRename)

numCellType <- case_when(ct == "CD14+TREM2+FOLR2-ABCG+ Lipid Associated Macrophages"~1,
                         ct == "CD14+IL1B+SELL+CD16- Antigen-presenting Inflammatory Monocyte-derived Macrophages"~2,
                         ct == "CD14+IL1B+SELL+S100A8+ Migrating Inflammatory Monocyte-derived Macrophages"~3,
                         ct == "CD14+TREM2-OLR1+NLRP3+ Inflammatory Foamy Macrophages"~4,
                         ct == "CD14+IL1B-TREM2-FOLR2+ Resident-like Macrophages"~5,
                         ct == "CD14+IL1B+SELL+MX1+ Interferon Activated Inflammatory Monocyte-derived Macrophages"~6,
                         ct == "CD14+-IL1B+SELL+CD16+ Migrating Inflammatory Monocyte-derived Macrophages"~7,
                         ct == "CD14+TNF+TREM2+FOLR2+ Inflammatory Resident-like Lipid Associated Macrophages"~8,
                         ct == "CD14+TREM2-OLR1+ABCA+ Foamy Macrophages"~9,
                         ct == "CD4+ T Cells"~10,
                         ct == "CD8+ T Cells"~11,
                         ct == "CD56-CD16+ NK Cells"~12,
                         ct == "CD79+ Class-switched Memory B Cells"~13,
                         ct == "FOXP3+ T Cells"~14,
                         ct == "CD56+KLRC1+ NK Cells"~15,
                         ct == "CD1C+ cDC1"~16,
                         ct == "CD34+ Endothelial Cells"~17,
                         ct == "ACTA2+ Smooth Muscle Cells"~18,
                         ct == "CD3+MKI67+ Proliferating T Cells"~19,
                         ct == "KIT+ Mast Cells"~20,
                         ct=="CLEC9A+ cDC2"~21)

newLabel <- paste0(numCellType, ": ", ct)
Idents(scRename) <- numCellType

plotCluster21 <- DimPlot(scRename,
                         reduction = "umap", 
                         label = TRUE, 
                         label.size = 6, 
                         label.box = TRUE, 
                         repel = TRUE, 
                         dims = c(1, 2),
                         order = sort(unique(numCellType), decreasing = T)) +
  theme_classic() +
  theme(axis.title = element_text(size = 15),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size = 15)) +
  guides(color = guide_legend(override.aes = list(size=4), ncol=1)) +
  scale_color_discrete(labels = unique(newLabel)[order(as.numeric(gsub(":.*", "", unique(newLabel))))])

tiff("output/figures/sc_plaque_cluster_21.tiff",
     width = 1280,
     height = 720,
     units = "px")

print(plotCluster21)

dev.off()
```

In order to assess whether clustering of certain cell types would reduce the complexity of the deconvolution, the 21 cell types will be reduced to 7 cell types.

```{r}
scCellTypes21 <- as.data.frame(Idents(scFinal)) %>%
  rename(Celltype = `Idents(scFinal)`)

unique(scCellTypes21$Celltype)

#7 cell types
cluster7 <- list(
  "Macrophage" = unique(str_subset(scCellTypes21$Celltype, "(?i)macrophage")),
  "T/NK cell" = unique(str_subset(scCellTypes21$Celltype, "((?i)\\sT\\scell)|((?i)NK cell)")),
  "Dendritic cell" = unique(str_subset(scCellTypes21$Celltype, "(cDC1)|(cDC2)")),
  "Memory B cell" = unique(str_subset(scCellTypes21$Celltype, "(?i)\\sB\\scell")),
  "Smooth muscle cell" = unique(str_subset(scCellTypes21$Celltype, "(?i)smooth")),
  "Endothelial cell" = unique(str_subset(scCellTypes21$Celltype, "(?i)endothelial")),
  "Mast cell" = unique(str_subset(scCellTypes21$Celltype, "(?i)mast"))
)

scCellTypes7 <- scCellTypes21

clType <- as.character(scCellTypes7$Celltype)

for (level in levels(scCellTypes7$Celltype)) {
  for (name in names(cluster7)) {
    if (level %in% unlist(cluster7[[name]])) {
      clType[clType %in% unlist(cluster7[[name]])] <- name
    }
  }
}

scCellTypes7$Celltype <- factor(clType, levels = names(cluster7))

unique(scCellTypes7$Celltype)

```

```{r}
ct7 <- case_when(ct %in% c("CD14+TREM2+FOLR2-ABCG+ Lipid Associated Macrophages",
                            "CD14+IL1B+SELL+CD16- Antigen-presenting Inflammatory Monocyte-derived Macrophages",
                            "CD14+IL1B+SELL+S100A8+ Migrating Inflammatory Monocyte-derived Macrophages",
                            "CD14+TREM2-OLR1+NLRP3+ Inflammatory Foamy Macrophages",
                            "CD14+IL1B-TREM2-FOLR2+ Resident-like Macrophages",
                            "CD14+IL1B+SELL+MX1+ Interferon Activated Inflammatory Monocyte-derived Macrophages",
                            "CD14+-IL1B+SELL+CD16+ Migrating Inflammatory Monocyte-derived Macrophages",
                            "CD14+TNF+TREM2+FOLR2+ Inflammatory Resident-like Lipid Associated Macrophages",
                            "CD14+TREM2-OLR1+ABCA+ Foamy Macrophages")~"Macrophage",
                  ct %in% c("CD4+ T Cells",
                            "CD8+ T Cells",
                            "CD56-CD16+ NK Cells",
                            "FOXP3+ T Cells",
                            "CD56+KLRC1+ NK Cells",
                            "CD3+MKI67+ Proliferating T Cells")~"T/NK cell",
                  ct %in% c("CD1C+ cDC1",
                            "CLEC9A+ cDC2")~"Dendritic cell",
                  ct=="CD79+ Class-switched Memory B Cells"~"Memory B cell", 
                  ct=="CD34+ Endothelial Cells"~"Endothelial cell", 
                  ct=="ACTA2+ Smooth Muscle Cells"~"Smooth muscle cell",
                  ct=="KIT+ Mast Cells"~"Mast cell")

numCellType7 <- case_when(ct7=="Macrophage"~1,
                           ct7=="T/NK cell"~2,
                           ct7=="Dendritic cell"~3,
                           ct7=="Memory B cell"~4,
                           ct7=="Endothelial cell"~5,
                           ct7=="Smooth muscle cell"~6,
                           ct7=="Mast cell"~7)

newLabel7 <- paste0(numCellType7, ": ", ct7)

scRename7 <- scFinal
Idents(scRename7) <- numCellType7

plotCluster7 <- DimPlot(scRename7,
                         reduction = "umap", 
                         label = TRUE, 
                         label.size = 6, 
                         label.box = TRUE, 
                         dims = c(1, 2),
                         order = sort(unique(numCellType7), decreasing = T)) +
  theme_classic() +
  theme(axis.title = element_text(size = 15),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size = 15)) +
  guides(color = guide_legend(override.aes = list(size=4), ncol=1)) +
  scale_color_discrete(labels = unique(newLabel7)[order(as.numeric(gsub(":.*", "", unique(newLabel7))))])

tiff("output/figures/sc_plaque_cluster_7.tiff",
     width = 720,
     height = 720,
     units = "px")

print(plotCluster7)

dev.off()
```

### Format files for Scaden

For the deconvolution method "Scaden", the input files need to be correctly formatted. For detailed instructions, see the [Scaden documentation page](https://scaden.readthedocs.io/en/latest/index.html).

```{r message=FALSE, warning=FALSE}
scCounts <- scFinal[["RNA"]]$counts %>%
  as.data.frame() %>%
  t()

#Check if the rownames of the two data frames are in the same order.
table(rownames(scCounts) == rownames(scCellTypes7))

write.table(scCellTypes7,
            file = "output/scaden_plaque/format_files/scaden_celltypes.txt",
            sep = "\t", row.names = FALSE)

write.table(scCounts,
            file = "output/scaden_plaque/format_files/scaden_counts.txt",
            sep = "\t", row.names = FALSE)
  
write.table(bulkFinal,
            file = "output/scaden_plaque/format_files/scaden_bulk_data.txt",
            sep = "\t")
```

## Neutrophils single cell RNA-sequencing data processing

As neutrophils are difficult to capture using the standard scRNA-seq workflow due to their fragility during isolation and low RNA content, among others, cellular deconvolution may provide an alternative method to capturing the neutrophil signal in bulk RNA-seq data.

### Importing files

```{r}
scNeutData <- read_rds("input/Alsaigh_scRNA/Alsaigh_2020.rds")
```

### Filter neutrophil single cell RNA-sequencing data

```{r}
DefaultAssay(scNeutData) <- "RNA"

toKeep <- intersect(Features(scNeutData), bulkFilter$symbol)

bulkNeutFinal <- bulkFilter %>%
  filter(symbol %in% toKeep) %>%
  as.data.frame() %>%
  column_to_rownames("symbol") %>%
  subset(select = -ensembl_gene_id)

countsNeut <- scNeutData[["RNA"]]$counts

countsKeep <- countsNeut[which(rownames(countsNeut) %in% toKeep),]

scNeutFilter <- subset(scNeutData, features = rownames(countsKeep))
```

Both the bulk RNA-seq and neutrophil scRNA-seq datasets contain 16053 genes.

```{r}
#Filter on the frequency of celltypes.
celltypes <- scNeutFilter@meta.data %>%
  select(SingleR.calls)

dfCellTypes <- as.data.frame(sort(table(celltypes)))

aboveThreshold <- as.character(factor(dfCellTypes$SingleR.calls[dfCellTypes$Freq >= 50]))

cellsKeep <- rownames(scNeutFilter@meta.data[scNeutFilter@meta.data$SingleR.calls %in% aboveThreshold, ])

scNeutFinal <- subset(scNeutFilter, cells = cellsKeep)
```

### Clustering

```{r}
scNeutCellTypes <- scNeutFinal@meta.data %>%
  select(SingleR.calls)

scCounts <- as.data.frame(scNeutFinal[["RNA"]]$counts) %>%
  t()

numCellType7 <- case_when(ct7=="Macrophage"~1,
                           ct7=="T/NK cell"~2,
                           ct7=="Dendritic cell"~3,
                           ct7=="Memory B cell"~4,
                           ct7=="Endothelial cell"~5,
                           ct7=="Smooth muscle cell"~6,
                           ct7=="Mast cell"~7)
```

### Format files for Scaden

```{r}
scCounts <- as.data.frame(scNeutFinal[["RNA"]]$counts) %>%
  t()

#Get the one where the cell types are clustered.
scNeutCellTypes <- scNeutFinal@meta.data %>%
  select(SingleR.calls)

#Check if the rownames of the two data frames are in the same order.
table(rownames(...) == rownames(...))

write.table(scCellTypes7,
            file = "output/scaden_plaque/format_files/scadenNeutrophil_celltypes.txt",
            sep = "\t", row.names = FALSE)

write.table(scCounts,
            file = "output/scaden_plaque/format_files/scadenNeutrophil_counts.txt",
            sep = "\t", row.names = FALSE)

#COPY BULK DATA TO NEW FOLDER AND RENAME IT TO HAVE scadenNeutrophil INCLUDED
```

## Session info

```{r}
devtools::session_info()
```
